### ISSUE TITLE
Documentation?

### ISSUE THREAD

**Issue by heviiguy** [eneiluj/integration_suitecrm#2]
Looks to be very promising. However, the sparse README and subsequent link to SuiteCRM hardly provides enough hand-holding to get this thing off of the ground for me. Is there any hope of having a few more installation hints?

**Comment by Jawn78** [comment]
Yeah, I am having issues. If I understand correctly it almost appears like it's pointing to the wrong api url. But without any documentation I don't really have the time to dig into the code.  This could be awesome!

**Comment by cpietras** [comment]
Hello,

In Admin/Connected Accounts, just put the URL https://domain.com with App ID and App Secret configured in SuiteCRM.

The pb is that it is working only if your SuiteCRM user is in local. If it is a LDAP account, it will not work. For example, local admin is working and my LDAP account not. Maybe a bad configuration from my side ?

**Comment by heviiguy** [comment]
Can't log in to SuiteCRM for the initial setup in Nextcloud: "Password is invalid"

I have no problem logging in to SuiteCRM directly when using the same credentials

**Comment by heviiguy** [comment]
Delving further into this, it appears that I must first setup a self-hosted OAuth2 Authorization Server. No thanks; it's becoming much more trouble than it's worth.

**Comment by heviiguy** [comment]
Arrggghhh... I can't stay way from this; It would be sooo good to be able to consolidate my reminders and notifications on Nextcloud.
Julien, if you could  just throw some more crumbs of hints at me so that I can get this thing working, I commit to writing a concise README so that others who are equally currently flummoxed can experience the functionality which this app promises.

**Comment by pgorod** [comment]
You guys are aware of this recent announcement, aren't you?

https://apps.nextcloud.com/apps/integration_suitecrm
https://suitecrm.com/salesagility-and-nextcloud-announce-suitecrm-integration

**Comment by eneiluj** [comment]
@heviiguy I'll write a proper documentation soon. Sorry about that.

Quick tutorial:
* a few requirements
    * your user must be local (as @DicDicus said), it won't work with LDAP users
    * it won't work if there is an SSO portal in front of your SuiteCRM
* there is a [small procedure](https://docs.suitecrm.com/developer/api/developer-setup-guide/json-api/#_generate_private_and_public_key_for_oauth2) to follow to generate private and public.key for OAUTH2 in SuiteCRM
* then create a new "password client" in your SuiteCRM admin settings, fill App ID and App secret in Nextcloud "Connected accounts" admin settings section
* then you should be able to log in from the "Connected accounts" user settings section

Any better?

**Comment by heviiguy** [comment]
Thanks, Julien. It's the small procedure that, I believe, is at the root of my problem. Whereas it seems that I've generated the necessary keys, I'm not sure how to verify if this is indeed true. Furthermore, in _which_ `config.php` must the  "oauth2_encryption_key" entry be edited? 

Yes, I'm quite aware of the integration announcement. It had me excited! It was a familiar feeling as I had felt the same when learning about the Nextcloud Talk developments. However, both had also resulted in frustration after realising that implementation is not so easy after all. Seems to be an effective model to whet user's appetites for premium paid support only for them to become dejected when realising that its cost is only realistic for large corporations. Oh well; I'll continue to take great benefit from this brilliant program and try to make sense of the enticing scraps!

**Comment by heviiguy** [comment]
I've not given up...
> * then create a new "password client" in your SuiteCRM admin settings, fill App ID and App secret in Nextcloud "Connected accounts" admin settings section

After having followed the instructions given by the [small procedure](https://docs.suitecrm.com/developer/api/developer-setup-guide/json-api/#_generate_private_and_public_key_for_oauth2), I found no evidence of such keys in my SuiteCRM instance (I looked at "OAuth Keys"). However, they clearly appear within the `{suitecrm.root}}/Api/V8/OAuth2` directory as **private.key** and **public.key**

I created  a new client at "OAuth2 Clients and Tokens". Assuming that the "App ID" is the OAuth2 Client ID, I have it.  However, where does one find the "App Secret"? There was nothing in the [small procedure](https://docs.suitecrm.com/developer/api/developer-setup-guide/json-api/#_generate_private_and_public_key_for_oauth2) where anything about a secret was mentioned.

**Comment by heviiguy** [comment]
So, waddling through this, I've created a new OAuth Consumer key by creating my own password and secret. I'm not sure of this is even necessary. In any case, neither this nor the actions described above resulted in a  OAuth2 "Token" (which, likewise, I'm not sure is required.  

>     * fill App ID and App secret in Nextcloud "Connected accounts" admin settings section

This one really perplexes me. Even if I had an App Secret, there is no option for this nor for App ID in "Connected accounts". The only available fields are:

- SuiteCRM instance address
- User name 
- Password

When entering the correct credentials, I'm stopped by: **_Failed: Invalid login/password_**

**Comment by heviiguy** [comment]
> no option for this nor for App ID in "Connected accounts"

Yes there is - if I go to the proper "Connected accounts" under **Administration**. Oops . Nevertheless, I'm still stuck; which "secret" are we to use?

**Comment by riker513** [comment]
I've got to add I'm having the same problem. While other integrations such as Gitlab connect successfully, I just can't get the SuiteCRM integration to work. Really spent some time on it reading up on the docs etc, but no success. nextcloud.log says that the API call returns the message that "OAuth2Clients module with id <id> is not found". I checked that what is given there as <id> is indeed the id I've been given after creating the password client in the admin section of SuiteCRM.
Thanks for the work and kind regards!

**Comment by beginner83510** [comment]
appID is the name you create in suiteCRM, secret is given

I get a javascript error when connection, even if javascript is enabled in browser

**Comment by heviiguy** [comment]
@beginner83510,
Thanks for your assistance. I hope that you have patience... 

I've named the OAuth2 Client as MySuiteCRM . Therefore,
- the **appID** is  "MySuiteCRM", right? 
- What significance does the  **ID** field of "MySuiteCRM" have?
- after selecting Edit, I renewed my secret by adding a fresh secret in the **Change Secret Field** 

Still no joy; "Password invalid"

Yes, I have a public and a private key at .../Api/V8/OAuth2/

**Comment by neonsoftware** [comment]
Hi @heviiguy, 

I admit that I have not finished the setup myself (nginx issues), please maintainers correct me if I'm wrong (?), but from the documentation ([suiteCrm oauth](https://docs.suitecrm.com/developer/api/developer-setup-guide/configure-authentication/), and other support pages) it seems that the three elements are, instead, as follows:

[see A, B, C in the screenshots]

A - Your "MySuiteCRM", you will not use it, is just to distinguish within SuiteCRM 
B - **Secret** : **you actually have to type it, something, yourself**, even if the UI seems to suggest otherwise, (i see that you changed it, but writing this down just for future record), see screenshot 
C - **appID**, or **ID**, is the generated 37a455df-14cf-9ea5-5da6

![immagine](https://user-images.githubusercontent.com/9502999/108788941-67544280-7579-11eb-9325-e320e9456eae.png)

Then

![immagine](https://user-images.githubusercontent.com/9502999/108789090-cca83380-7579-11eb-8674-3185a7fb84f8.png)

So finally
![immagine](https://user-images.githubusercontent.com/9502999/108789772-4a207380-757b-11eb-9236-e629bf1d7216.png)

Let us know the result using this configuration. 👍🏼

**Comment by neonsoftware** [comment]
@Jawn78 for the 404, is your issue solved ? I have seen some pages in that regards:
If you use Apache:
- [this step in the documentation about mod_rewrite and AllowOverride](https://docs.suitecrm.com/developer/api/developer-setup-guide/json-api/#_verify_if_rewrite_module_is_installed_and_activated)
- [many support tickets mentioning .htaccess](https://community.suitecrm.com/t/api-access-token-shows-404-file-not-found/69910/2)

If instead you use nginx or other servers, it might be a redirection issue, as an equivalent of Apache's mod_rewrite should be manually setup, but have not yet found one to work yet. I currently have the same 404 issue with nginx. @eneiluj @pgorod @DicDicus  any suggestion for an nginx config maybe ? Thanks in advance

**Comment by heviiguy** [comment]
Thank you very much, @neonsoftware . It's clear that you took extraordinary effort and time to help. This is appreciated!
What you so effectively illustrated is exactly how I had eventually populated my fields. Nevertheless, I still can't get in...

Although I thought that my .config was configured properly, I'll have yet another look.

**Comment by Ozzie76** [comment]
Does anybody know if the public and private keys are upgrade persistent in a Docker version of SuiteCRM?
FYI, I'm using the Bitnami docker container on a Synology NAS. For some reason the container crashes when I try to terminal into it. I hope somebody with more Docker experience can prevent me chasing my tail on this.

**Comment by warden-ltd** [comment]
How about
SuiteCRM OAuth error : cURL error 60: SSL certificate problem: unable to get local issuer certificate (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://ourdomain.com/Api/access_token

public and private keys generated and save in default localisation, certificates are correct, ca-ceritficates is installed and updated, i`ve even tried cacert.pem and curl.ca path solution in php.ini but still get error 60. Any ideas?

**Comment by AreYouLoco** [comment]
I have found out that you will need OAuth2 server. So basically you need to install composer and run `sudo -u www-data composer require league/oauth2-server` in /var/www/suite or wherever you keep your SuiteCRM. Then it installs OAuth2 server but then how to proceed?

**Comment by brendan-pike** [comment]
Hi Guys, I've been stuck trying to get it to work also. I suspect my issue is actually OAuth2 on SuiteCRM side, however I'm posting here because I feel its still part of the lack of clear documentation since changes need to be made to the default SuiteCRM install to support this.  If you have a chance could you look at my support post I made here please; https://community.suitecrm.com/t/unable-to-get-oauth2-working-with-nextcloud-app/83847

**Comment by Ozzie76** [comment]
@brendan-pike your message triggered me to try again. Haven't touched it in a while, but had success tonight. My key points:
1. add to the NextCloud config.php the line: ‘allow_local_remote_servers’ => true,
2. in the NextCloud settings, enter your URL in the line "SuiteCRM instance address" like in neonsoftware's example above. The system will add the following to the URL: "/Api/access_token". I had doubles in there, so make sure you point to the base folder of SuiteCRM and leave out the trailing slash ("/").
Kudo's to this everyone here and SteffnVo over at the NextCloud forum for pointing me in the right direction. https://help.nextcloud.com/t/suitecrm-integration/101794
